src/subdomains/core/payment-link/services/payment-link-payment.service.ts

```diff
@@ -184,11 +185,16 @@ export class PaymentLinkPaymentService {
     if (pendingPayment)
       throw new ConflictException('There is already a pending payment for the specified payment link');
 
+    const allPayments = await this.paymentLinkPaymentRepo.find({
+      where: { link: { uniqueId: paymentLink.uniqueId } },
+    });
+
+    if (paymentLink.mode === PaymentLinkMode.SINGLE) {
+      if (allPayments.length > 0) throw new ConflictException('Single payment link can only have one payment');
+    }
+
     if (dto.externalId) {
-      const exists = await this.paymentLinkPaymentRepo.existsBy({
-        externalId: dto.externalId,
-        link: { id: paymentLink.id },
-      });
+      const exists = allPayments.some((p) => p.externalId === dto.externalId);
       if (exists) throw new ConflictException('Payment already exists');

```

Comment:
I think I would split this in two DB requests. And only if payment link mode is SINGLE, check if there are already payments with an `exists` call. Loading all payments for a link might result in a huge amount of data.